Use env_spec in algos

naeioi opened the pull request:
This PR replaces env with env_spec in algorithm constructor.

This closes #513 .

ryanjulian commented on garage/experiment/local_tf_runner.py:
```diff
@@ -120,19 +120,20 @@ def setup(self, algo, env, sampler_cls=None, sampler_args=None):
-                if v.name.split(':')[0] in str(
-                    self.sess.run(tf.report_uninitialized_variables()))
-            ]))
+        with tf.name_scope("auto-initialized"):
```
hmm what prompted this change?

CatherineSue replied:
The current tf graph looks like this: (many report_uninitialized_variables)
![image](https://user-images.githubusercontent.com/8605658/53993779-2db47380-40e5-11e9-82db-1a86d0270c29.png)

ryanjulian replied:
ah i see. can you change the scope name to `'initialize_tf_vars'`?

That will make it easy for someone find what code produced this when they see it in TensorBoard.

ryanjulian replied:
also, can  you show the "after" TensorBoard corresponding to Chang's "before"?

naeioi replied:
<img src="https://user-images.githubusercontent.com/8409272/54252202-2c30e400-4507-11e9-8e0d-df289ea41ca8.png" height=500px>

ryanjulian replied:
looks great.

ryanjulian commented on garage/tf/algos/ddpg.py:
```diff
@@ -47,7 +47,7 @@ def __init__(self,
 
         Args:
-            env(): Environment.
+            env_spec(): Environment specification.
```
does't this have a type?

naeioi replied:
It is `akro.Space`?

ryanjulian replied:
uhh i think it's https://github.com/rlworkgroup/garage/blob/master/garage/envs/env_spec.py for now.

ryanjulian commented on garage/tf/samplers/on_policy_vectorized_sampler.py:
```diff
@@ -13,25 +13,25 @@
 
-        if getattr(self.algo.env, 'vectorized', False):
-            self.vec_env = self.algo.env.vec_env_executor(
+        if getattr(self.env, 'vectorized', False):
```
im not sure this case even exists anymore. you can probably delete it.

naeioi commented:
@ryanjulian Because the snapshot is taken by the algorithm and I removed env from the pickled file, it turns out that there's no way to save env.  My idea is that the saving env is not very important so I just removed it. 

Also, I doubt the experiement resuming really worked previously because `algo.train()` will randomized all variables. It's neither working currently because `run_experiment()` still calls `algo.train()` which is not valid anymore. https://github.com/rlworkgroup/garage/blob/1c299fd56b90f641321cfc0482eca6770256f49c/scripts/run_experiment.py#L174-L178

I plan to resolve this issue together with #516 .

ryanjulian commented on garage/experiment/local_tf_runner.py:
```diff
@@ -172,7 +173,7 @@ def save_snapshot(self, itr, paths=None):
-        params = self.algo.get_itr_snapshot(itr, paths)
+        params = self.algo.get_itr_snapshot(itr)
         if paths:
             params["paths"] = paths
```
untested

ryanjulian commented:
@naeioi  it's quite important to serialize the environment, and we can't really merge this without saving the environment.

Look at https://github.com/rlworkgroup/garage/blob/master/examples/sim_policy.py -- this PR breaks it. 

TODO: We should add a test to make sure the snapshotter produces complete .pkl files, so that this PR would fail in the CI immediately.

Reasons:
1. There's no reason `env` can't have meaningful parameters which are saved/restored during pickle/unpickle -- for instance, what if I had a PointMass environment with specifiable mass? Not all envs are just strings from openai/gym. The only interface we impose on `env` that (1) `env == pickle.loads(pickle.dumps(env))` and (2) `isinstance(env, gym.Env)`
2. A `.pkl` file should be a total record of an experiment -- it should have everything I need to evaluate an agent in the environment, continue training the agent in the environment, etc. removing `Env` makes that untrue -- i would have to look at the launcher and be careful to recreate the environment. consider how hard this would be if I decided to procedurally generate my environments!
3. Perhaps (2) isn't the best way to save experiment state, but right now it's what we have.

naeioi commented:
@ryanjulian That makes sense. I added a test to verify the presence and integrity of both environment and policy in snapshot. As a workaround of algo not having access to the actual environment, the environment is now saved in `save_snapshot()` in the runner.

ryanjulian commented:
awesome!

ryanjulian commented on tests/garage/experiment/test_snapshot.py:
```diff
@@ -0,0 +1,70 @@
+        with LocalRunner():
+            snapshot = joblib.load(osp.join(self.log_dir, 'params.pkl'))
+            env = snapshot['env']
+            policy = snapshot['policy']
```
perhaps

```python
assert env
assert policy
```
Just so that any error leads me to this line, instead of some line inside `rollout()`

naeioi replied:
Good point! Fixed.

ryanjulian commented on garage/experiment/local_tf_runner.py:
```diff
@@ -171,8 +172,11 @@ def save_snapshot(self, itr, paths=None):
         logger.log("Saving snapshot...")
-        params = self.algo.get_itr_snapshot(itr, paths)
+        params = self.algo.get_itr_snapshot(itr)
+        params['env'] = self.env
```
hmm can you do this in `save_snapshot` instead? that covers the case where an environment has evolving state.

it would be good to test for that case too, since it would be incredibly hard to detect/debug...

naeioi replied:
Are you saying that I should get the snapshot of env from sampler because the copy of env in the runner is not the actual copy used for sampling? 

This leads to another problem: What to save when there are more than one env in the sampler. We can save all envs, but we dont't actually have a mechanism to restore them in sampler. Or we pick one env and save it, and that will be compatible to the current things we save in snapshot.

ryanjulian replied:
hmm, this presents an interesting problem gap in the interface, but i'm not sure it's one we need to solve right now.

for now i think it's sufficient to make sure there's a copy of `env` in every .pkl file. we can try to handle statefulness later if there's a need.

naeioi replied:
OK. Let's keep an eye on it.

ryanjulian commented on garage/tf/algos/ddpg.py:
```diff
@@ -197,7 +197,7 @@ def train_once(self, itr, paths):
             logger.log("Training finished")
             logger.log("Saving snapshot #{}".format(int(epoch)))
-            params = self.get_itr_snapshot(epoch, paths)
+            params = self.get_itr_snapshot(epoch)
```
is `params` even used?

naeioi replied:
No. And the `save_itr_params` is not necessary. I will delete it.

ryanjulian commented on tests/garage/experiment/test_snapshot.py:
```diff
@@ -0,0 +1,75 @@
+        # Read snapshot from self.log_dir
+        # Test the presence and integrity of policy and env
+        with LocalRunner():
+            snapshot = joblib.load(osp.join(self.log_dir, 'params.pkl'))
```
please add a test case which makes sure that the environment is present in the pkl file generated by every iteration (e.g. 2 or 3)

naeioi replied:
Added.

ryanjulian commented on tests/garage/experiment/test_snapshot.py:
```diff
@@ -0,0 +1,75 @@
+
+            env.close()
+
+        self.tearDown()
```
hmm this is pretty confusing to see in a test, and makes you vulnerable to arbitary changes in TfGraphTestCase. I think you should just define local helpers for this.

naeioi replied:
Addressed.

CatherineSue commented on examples/tf/her_ddpg_fetchreach.py:
```diff
@@ -51,14 +51,14 @@ def run_task(*_):
-            plot=False,
             target_update_tau=0.05,
+            n_epoch_cycles=20,
             max_path_length=100,
```
I thought this line is deleted in off policy algorithm?

naeioi replied:
Actually not. I keep `n_epoch_cycles` so that an algorithm can compute epoch by `iteration/n_epoch_cycles`. This is a little bit awkward but required in DDPG.

CatherineSue replied:
I don't quite understand. I meant is `max_path_length=100` not necessary here?

ryanjulian commented:
bump. are there any hard blockers on this PR?

naeioi commented:
I will have this merged after rebase.

naeioi commented:
I disable two pylint checks. 
- `not-context-manager check`
Pylint has long been having bug on checking context manager.
See https://github.com/PyCQA/pylint/issues/782.

- `c-extension-no-member check`
Some C modules don't report their functions and variable in python. In our case, baselines is using mpi4py which makes pylint complain.
See https://travis-ci.com/rlworkgroup/garage/builds/105359651#L690.
